### ISSUE TITLE
Is UDP proxying supported

### ISSUE THREAD

**Issue by Overdrivr** [dokku/dokku#2767]
Description of problem:

I am trying to send metrics from a dokku app to a StatsD daemon without success so far.
The [StatsD daemon](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/statsd) expects metrics to arrive on UDP:8125.

I tried naively the following command to proxy the udp port, without success.

```
dokku proxy:ports-add myapp udp:8125:8125
```

I've seen written [here](https://github.com/dokku/dokku/issues/1629#issuecomment-173828473) that Dokku does not support UDP proxying between the machine and container. Is this still true ? It would clearly explain why metrics never reach StatsD.

I haven't seen mentioned in the docs on [proxy management](http://dokku.viewdocs.io/dokku~v0.9.4/advanced-usage/proxy-management/) that UDP is not supported. Might be worth adding.

Also, the proxy:ports-add command could throw an error if trying to proxy with an unsupported scheme. Would avoid others a few hours of scratching their heads.

* If UDP is not an option, how do people go around monitoring dokku apps with StatsD ?  
* Otherwise, how can I enable proxying udp requests out of the container ?

PS: I initially asked on SO without answers so far, and there's a [100 rep bounty question](http://stackoverflow.com/questions/43822074/telegraf-daemon-not-receiving-metrics-from-app-deployed-with-dokku) ups for grabs. 
```
-----> uname: Linux XXXXXX 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
-----> memory:
              total        used        free      shared  buff/cache   available
Mem:           1952         249         236          20        1465        1432
Swap:             0           0           0
-----> docker version:
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64
 Experimental: false
-----> docker daemon info:
Containers: 15
 Running: 1
 Paused: 0
 Stopped: 14
Images: 24
Server Version: 17.04.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 74
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary:
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-62-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.907GiB
Name: XXXXXXXXXXXXXXXXXXX
ID: XXXXXXXXXXXXXXXXXXXXXXXXXX
Docker Root Dir: /var/lib/docker
Debug Mode (client): true
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
-----> sigil version: 0.4.0
-----> herokuish version:
herokuish: 0.3.27
buildpacks:
  heroku-buildpack-multi     v1.0.0
  heroku-buildpack-ruby      v155
  heroku-buildpack-nodejs    v98
  heroku-buildpack-clojure   v76
  heroku-buildpack-python    v99
  heroku-buildpack-java      v51
  heroku-buildpack-gradle    v21
  heroku-buildpack-grails    v21
  heroku-buildpack-scala     v76
  heroku-buildpack-play      v26
  heroku-buildpack-php       v120
  heroku-buildpack-go        v62
  heroku-buildpack-erlang    fa17af9
  buildpack-nginx            v8
-----> dokku version: 0.9.2
-----> dokku plugins:
 !     Deprecated: Please use plugin:list
plugn: 0.3.0
  00_dokku-standard    0.9.2 enabled    dokku core standard plugin
  20_events            0.9.2 enabled    dokku core events logging plugin
  apps                 0.9.2 enabled    dokku core apps plugin
  build-env            0.9.2 enabled    dokku core build-env plugin
  certs                0.9.2 enabled    dokku core certificate management plugin
  checks               0.9.2 enabled    dokku core checks plugin
  common               0.9.2 enabled    dokku core common plugin
  config               0.9.2 enabled    dokku core config plugin
  docker-options       0.9.2 enabled    dokku core docker-options plugin
  domains              0.9.2 enabled    dokku core domains plugin
  enter                0.9.2 enabled    dokku core enter plugin
  git                  0.9.2 enabled    dokku core git plugin
  letsencrypt          0.8.6 enabled    Automated installation of let's encrypt TLS certificates
  logs                 0.9.2 enabled    dokku core logs plugin
  named-containers     0.9.2 enabled    dokku core named containers plugin
  nginx-vhosts         0.9.2 enabled    dokku core nginx-vhosts plugin
  plugin               0.9.2 enabled    dokku core plugin plugin
  proxy                0.9.2 enabled    dokku core proxy plugin
  ps                   0.9.2 enabled    dokku core ps plugin
  repo                 0.9.2 enabled    dokku core repo plugin
  shell                0.9.2 enabled    dokku core shell plugin
  ssh-keys             0.9.2 enabled    dokku core ssh-keys plugin
  storage              0.9.2 enabled    dokku core storage plugin
  tags                 0.9.2 enabled    dokku core tags plugin
  tar                  0.9.2 enabled    dokku core tar plugin

```

**Comment by josegonzalez** [comment]
This is correct, we do not support udp routing with nginx at this time. Why? Because the common use case for dokku is for web application deployment. Web applications typically do not use udp or tcp sockets, and instead listen on http.

Related issue #2662. We also have the [dokku-graphite-grafana](https://github.com/dokku/dokku-graphite-grafana) plugin, but it has the same issue regarding proxying udp due to missing functionality in the ambassador approach.

Because this isn't a necessity for most web applications, it won't land without a pull request from the community as I am not going to prioritize it's development. However, if you are interested in sponsoring it's development, feel free to contact me directly via irc or slack (I am `savant` on both).

**Comment by josegonzalez** [comment]
As well, it is up to the proxy implementation as to decide what schemes are supported. At the bottom of the page, we should how the schema works and note that `http` and `https` are the supported schemes for the built-in [nginx-vhosts plugin](http://dokku.viewdocs.io/dokku~v0.9.4/advanced-usage/proxy-management/#proxy-port-scheme).

**Comment by Overdrivr** [comment]
@josegonzalez But is there no workaround ? Like deploying with a Dockerfile where port 5000 is exposed for HTTP/HTTPS and 8125 for UDP. If nginx is only configured to proxy only port 5000, will the 8125 port be accessible ?

**Comment by josegonzalez** [comment]
You can turn off checks - disabling zero downtime - and then expose the udp port using the docker-options plugin for now.

**Comment by Overdrivr** [comment]
Thanks for your answer. I tried it without success, am I missing something here ? I ran `dokku ps:restart myapp` after disabling zero downtime and setting docker-options.

```
dokku docker-options:report
=====> myapp docker options information
       Docker options build:
       Docker options deploy: --restart=on-failure:10
       Docker options run:  -p 8125:8125/udp
```
```
dokku checks:report
=====> myapp checks information
       Checks disabled list: _all_
       Checks skipped list: none
```
```
dokku proxy:report
=====> myapp proxy information
       Proxy enabled:       true
       Proxy type:          nginx
       Proxy port map:      http:80:5000 https:443:5000
```

**Comment by josegonzalez** [comment]
Remove the `run` docker-option you've set and very carefully read about the [different docker options phases](http://dokku.viewdocs.io/dokku/advanced-usage/docker-options/#about-dokku-phases).

**Comment by Overdrivr** [comment]
Sorry about that, got confused about the stages name.

After fixing this, the solution you proposed is not working anyway, because it makes docker listen on port 8125 while there is already the statsd daemon listening on it. So deployment fails.

However, the good news is that it can be done very easily out of the box, without anything to do in Dokku.
Since the UDP payloads are sent from the docker container to a daemon that runs on the host, there is no port to open.

The only thing to do is to make sure UDP payloads are not sent to `localhost:8125` (in which case this is the container's localhost and cannot reach the host) but on the Docker gateway url, usually `172.17.0.1`. This way, they are relayed by Docker on the host's localhost url. And the daemon receives them properly. I just tested this successfully. There are other options than using the gateway url, see [here](http://stackoverflow.com/questions/43961530/by-default-can-a-docker-container-call-hosts-localhost-udp).

The good news is that zero downtime deploy can be left enabled, and there is no need to add a docker-options. 🍾 

Thanks again for helping me out on this issue, I've learned a great deal in the process. Cheers
